File: test/execution_context/f_execution_context_test.dart
// Copyright 2023, the Freemework.ORG project authors. Please see the AUTHORS
// file for details. All rights reserved. Use of this source code is governed
// by a BSD-style license that can be found in the LICENSE file.

import "package:collection/collection.dart" show DeepCollectionEquality;

import "package:test/test.dart";

import "package:freemework/freemework.dart";

class _TestExecutionContext extends FExecutionContext {
  const _TestExecutionContext(FExecutionContext prevContext)
      : super(prevContext);
}

void main() {
  group("FExecutionContext tests", () {
    setUp(() {});

    tearDown(() {});

    test("should be instantable without inner errors", () {
      FExceptionAggregate([]); // no expection expected
    });

    test("Empty execution context should NOT have prevContext", () {
      final FExecutionContext emptyCtx =
          FExecutionContext.emptyExecutionContext;
      expect(emptyCtx.prevContext, isNull);
    });

    test("Cancellation execution context should be resolved on head of chain",
        () {
      final FExecutionContext emptyCtx =
          FExecutionContext.emptyExecutionContext;
      final FExecutionContext cancellationCtx =
          FCancellationExecutionContext(emptyCtx, FCancellationToken.dummy);

      final FCancellationExecutionElement element =
          FCancellationExecutionContext.of(cancellationCtx);
      expect(element.owner, same(cancellationCtx));
      expect(element.cancellationToken, same(FCancellationToken.dummy));
    });

    test("Cancellation execution context should be resolved on chain", () {
      final FExecutionContext emptyCtx =
          FExecutionContext.emptyExecutionContext;
      final FExecutionContext cancellationCtx =
          FCancellationExecutionContext(emptyCtx, FCancellationToken.dummy);
      final stubCtx = _TestExecutionContext(cancellationCtx);

      final FCancellationExecutionElement element =
          FCancellationExecutionContext.of(stubCtx);
      expect(element.owner, same(cancellationCtx));
      expect(element.cancellationToken, same(FCancellationToken.dummy));
    });

    test("Cancellation execution context should aggregate tokens", () {
      final FExecutionContext emptyCtx =
          FExecutionContext.emptyExecutionContext;

      final FCancellationTokenSource cts1 = FCancellationTokenSourceManual();
      final FCancellationTokenSource cts2 = FCancellationTokenSourceManual();

      final FCancellationExecutionContext cancellationCtx1 =
          FCancellationExecutionContext(emptyCtx, cts1.token);
      final FCancellationExecutionContext cancellationCtx2 =
          FCancellationExecutionContext(cancellationCtx1, cts2.token, true);
      final stubCtx = _TestExecutionContext(cancellationCtx2);

      final FCancellationExecutionElement element =
          FCancellationExecutionContext.of(stubCtx);

      expect(element.owner, same(cancellationCtx2));
      expect(element.cancellationToken, isNot(FCancellationToken.dummy));
      expect(element.cancellationToken, isNot(FCancellationToken.dummy));
      expect(element.cancellationToken, isA<FCancellationTokenAggregated>());
    });

    test(
        "LoggerProperties execution context should be resolved on head of chain",
        () {
      final FLoggerLabels loggerLabels = {};

      final FExecutionContext emptyCtx =
          FExecutionContext.emptyExecutionContext;
      final FExecutionContext loggerCtx =
          FLoggerLabelsExecutionContext(emptyCtx, loggerLabels);

      final FLoggerLabelsExecutionElement? element =
          FLoggerLabelsExecutionContext.of(loggerCtx);
      expect(element, isNotNull);
      expect(element!.owner, same(loggerCtx));
      expect(
        DeepCollectionEquality().equals(
          element.loggerLabels,
          loggerLabels,
        ),
        true,
      );
    });

    // test("LoggerProperties execution context should be resolved on chain", () {
    //   final loggerLabels = <FLoggerProperty>[];

    //   final FExecutionContext emptyCtx = FExecutionContext.Empty;
    //   final FExecutionContext loggerCtx =
    //       FExecutionContextLoggerProperties(emptyCtx, loggerLabels);
    //   final stubCtx = _TestExecutionContext(loggerCtx);

    //   final FExecutionElementLoggerProperties element =
    //       FExecutionContextLoggerProperties.of(stubCtx);
    //   expect(element.owner, same(loggerCtx));
    //   expect(
    //     ListEquality<FLoggerProperty>().equals(
    //       element.loggerLabels.toList(),
    //       loggerLabels,
    //     ),
    //     true,
    //   );
    // });

    // test(
    //     "LoggerProperties execution context should instantiate from logger context",
    //     () {
    //   final prop1 = FLoggerProperty("data", "41");
    //   final prop2 = FLoggerProperty("data", "42");

    //   final loggerLabels = <FLoggerProperty>[prop1];

    //   final FExecutionContext emptyCtx = FExecutionContext.Empty;
    //   final FExecutionContext loggerCtx =
    //       FExecutionContextLoggerProperties(emptyCtx, loggerLabels);
    //   final FExecutionContext loggerWithPropertiesCtx =
    //       FExecutionContextLoggerProperties(loggerCtx, [
    //     prop2,
    //   ]);
    //   final stubCtx = _TestExecutionContext(loggerWithPropertiesCtx);

    //   final FExecutionElementLoggerProperties element =
    //       FExecutionContextLoggerProperties.of(stubCtx);
    //   expect(element.owner, same(loggerWithPropertiesCtx));
    //   expect(
    //     ListEquality<FLoggerProperty>().equals(
    //       element.loggerLabels.toList(),
    //       [
    //         prop2,
    //         prop1,
    //       ],
    //     ),
    //     true,
    //   );
    // });
  });
}

File: test/execution_context/f_execution_context.test.ts
import {
	FCancellationTokenAggregated, FCancellationTokenSource, FCancellationTokenSourceManual,
	FCancellationExecutionContext, FCancellationExecutionElement, FCancellationToken,
	FExecutionContext, FExecutionContextBase,
	FLoggerLabelsExecutionContext,
	FLoggerPropertiesExecutionElement,
	FLoggerLabels
} from "../../src";

import { assert } from "chai";

describe("FExecutionContext test", function () {
	it("Empty execution context should NOT have prevContext", function () {
		const emptyCtx: FExecutionContext = FExecutionContext.Empty;
		assert.isNull(emptyCtx.prevContext);
	});

	it("Cancellation execution context should be resolved on head of chain", function () {
		const emptyCtx: FExecutionContext = FExecutionContext.Empty;
		const cancellationCtx: FExecutionContext = new FCancellationExecutionContext(emptyCtx, FCancellationToken.Dummy);

		const element: FCancellationExecutionElement = FCancellationExecutionContext.of(cancellationCtx);
		assert.strictEqual(element.owner, cancellationCtx);
		assert.strictEqual(element.cancellationToken, FCancellationToken.Dummy);
	});

	it("Cancellation execution context should be resolved on chain", function () {
		const emptyCtx: FExecutionContext = FExecutionContext.Empty;
		const cancellationCtx: FExecutionContext = new FCancellationExecutionContext(emptyCtx, FCancellationToken.Dummy);
		const stubCtx = new StubExecutionContext(cancellationCtx);

		const element: FCancellationExecutionElement = FCancellationExecutionContext.of(stubCtx);
		assert.strictEqual(element.owner, cancellationCtx);
		assert.strictEqual(element.cancellationToken, FCancellationToken.Dummy);
	});

	it("Cancellation execution context should aggregate tokens", function () {
		const emptyCtx: FExecutionContext = FExecutionContext.Empty;

		const cts1: FCancellationTokenSource = new FCancellationTokenSourceManual();
		const cts2: FCancellationTokenSource = new FCancellationTokenSourceManual();

		const cancellationCtx1: FExecutionContext = new FCancellationExecutionContext(emptyCtx, cts1.token);
		const cancellationCtx2: FExecutionContext = new FCancellationExecutionContext(cancellationCtx1, cts2.token, true);
		const stubCtx = new StubExecutionContext(cancellationCtx2);

		const element: FCancellationExecutionElement = FCancellationExecutionContext.of(stubCtx);
		assert.strictEqual(element.owner, cancellationCtx2);
		assert.notStrictEqual(element.cancellationToken, cts1.token);
		assert.notStrictEqual(element.cancellationToken, cts2.token);
		assert.instanceOf(element.cancellationToken, FCancellationTokenAggregated);
	});

	it("Logger execution context should be resolved on head of chain", function () {
		const emptyCtx: FExecutionContext = FExecutionContext.Empty;
		const loggerCtx: FExecutionContext = new FLoggerLabelsExecutionContext(emptyCtx, { name: "test", value: "42" });

		const element: FLoggerPropertiesExecutionElement = FLoggerLabelsExecutionContext.of(loggerCtx)!;
		assert.isNotNull(element);
		assert.strictEqual(element.owner, loggerCtx);
		assert.strictEqual(element.loggerLabels.name, "test");
		assert.strictEqual(element.loggerLabels.value, "42");
	});

	it("Logger execution context should be resolved on chain", function () {
		const emptyCtx: FExecutionContext = FExecutionContext.Empty;
		const loggerCtx1: FExecutionContext = new FLoggerLabelsExecutionContext(emptyCtx, { name: "test", value: "42" });
		const loggerCtx2: FExecutionContext = new FLoggerLabelsExecutionContext(loggerCtx1, { name: "test", value: "43" });
		const stubCtx = new StubExecutionContext(loggerCtx2);

		const element: FLoggerPropertiesExecutionElement = FLoggerLabelsExecutionContext.of(stubCtx)!;
		assert.isNotNull(element);
		assert.strictEqual(element.owner, loggerCtx2);
		assert.deepEqual({ ...element.loggerLabels }, { name: "test", value: "42" });
	});
});

class StubExecutionContext extends FExecutionContextBase {
}
